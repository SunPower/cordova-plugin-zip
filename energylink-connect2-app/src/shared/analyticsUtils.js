import {
  __,
  compose,
  curry,
  gt,
  isEmpty,
  join,
  lensIndex,
  map,
  not,
  over,
  props,
  propSatisfies,
  toLower,
  toUpper,
  when
} from 'ramda'

export const isNotEmpty = compose(not, isEmpty)

export const capitalizeWord = when(
  propSatisfies(gt(__, 0), 'length'),
  compose(join(''), over(lensIndex(0), toUpper), toLower)
)

const stringValueOrEmptyString = stringValue => stringValue || ''

const getStringProps = curry(compose(map(stringValueOrEmptyString), props))

export const getUserProfile = getStringProps([
  'uniqueId',
  'firstName',
  'lastName',
  'email',
  'dealerName',
  'recordType'
])
