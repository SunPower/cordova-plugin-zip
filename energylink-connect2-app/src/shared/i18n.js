import { useMemo } from 'react'
import { useSelector } from 'react-redux'

import * as languages from './locales'
import * as localizedComponents from './localizedComponents'

const getTranslate = translations => (key, ...params) => {
  let value = translations[key] || key
  if (params) {
    params.forEach((p, idx) => {
      value = value.replace(`{${idx}}`, p)
    })
  }
  return value
}

export const useI18n = () => {
  const language = useSelector(state => state.language)
  const { locale = 'en' } = language
  return useMemo(() => getTranslate(languages[locale]), [locale])
}

export const useI18nComponent = () => {
  const language = useSelector(state => state.language)
  const { locale = 'en' } = language || {}

  return useMemo(
    () => componentName => {
      const path = `${locale}_${componentName}`
      const Component = localizedComponents[path] || {}
      if (typeof Component !== 'function') {
        throw new Error(`Component not found: ${path}`)
      }
      return Component()
    },
    [locale]
  )
}
