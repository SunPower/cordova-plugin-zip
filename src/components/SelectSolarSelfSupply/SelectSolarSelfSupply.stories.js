import React from 'react'

import { storiesOf } from '@storybook/react'

import SelectSolarSelfSupply from './index'

storiesOf('SelectSolarSelfSupply', module).add('Simple', () => (
  <SelectSolarSelfSupply />
))
