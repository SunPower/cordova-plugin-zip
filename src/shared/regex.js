export const PASSWORD_REGEXP = /^(?=.*[a-z])(?=.*\d)[A-Za-z\d]{8,}$/
export const EMAIL_REGEXP = /^\S+@\S+\.\S+$/
export const PHONE_REGEXP = /^(1\s?)?((([0-9]{3}))|[0-9]{3})[\s-]?[0-9]{3}[\s-]?[0-9]{4}$/
