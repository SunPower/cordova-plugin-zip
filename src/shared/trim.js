export const trimObject = obj => {
  return Object.entries(obj)
    .map(([k, v]) => ({
      [k]: typeof v === 'string' || v instanceof String ? v.trim() : v
    }))
    .reduce((acc, kv) => ({ ...acc, ...kv }), {})
}
