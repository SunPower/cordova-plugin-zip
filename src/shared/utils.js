export const either = (condition, whenTrue, whenFalse) =>
  condition ? whenTrue : whenFalse
