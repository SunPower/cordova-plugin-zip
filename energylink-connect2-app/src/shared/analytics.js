import appVersion from '../macros/appVersion.macro'
import { capitalizeWord, getUserProfile } from 'shared/analyticsUtils'
import { MIXPANEL_EVENT_QUEUED } from 'state/actions/analytics'
import { getAppleDeviceFamily } from 'shared/appleDevicesTable'
import { prop } from 'ramda'

const getAppAndDeviceProperties = () => {
  const { device } = window
  const { model, version, platform, manufacturer } = device

  let deviceModel
  let deviceBrowserVersion

  if (platform === 'iOS') {
    deviceModel = getAppleDeviceFamily(model)
    deviceBrowserVersion = version
  } else {
    deviceModel = `${manufacturer}, ${model}`
  }

  return {
    'App Build': appVersion(),
    'App Flavor': process.env.REACT_APP_FLAVOR,
    'OS Version': version,
    ...(deviceBrowserVersion ? { $browser_version: deviceBrowserVersion } : {}),
    $device: deviceModel
  }
}

export const loggedIn = user => {
  const { mixpanel } = window
  const {
    uniqueId: userId,
    firstName,
    lastName,
    email,
    dealerName,
    recordType: dealerType
  } = getUserProfile(user)
  const appAndDeviceProperties = getAppAndDeviceProperties()

  mixpanel.register(appAndDeviceProperties)
  mixpanel.identify(userId)
  mixpanel.people.set(appAndDeviceProperties)
  mixpanel.people.set({
    $first_name: firstName,
    $last_name: lastName,
    $email: email,
    'Dealer Name': dealerName,
    'Dealer Type': capitalizeWord(dealerType)
  })
  mixpanel.track('Login', { Success: true })

  return MIXPANEL_EVENT_QUEUED('Login - Success')
}

export const loginFailed = () => {
  const { mixpanel } = window
  const appAndDeviceProperties = getAppAndDeviceProperties()

  mixpanel.register(appAndDeviceProperties)
  mixpanel.track('Login', { Success: false })

  return MIXPANEL_EVENT_QUEUED('Login - Failed')
}

export const scanPVS = ({ pvsSN, entryMethod }, event) => {
  const { mixpanel } = window

  mixpanel.register({ 'PVS SN': pvsSN })
  mixpanel.track('Scan PVS Tag', {
    'Entry Method': entryMethod
  })

  return MIXPANEL_EVENT_QUEUED(event)
}

export const siteNotFound = ({
  uniqueId: userId,
  firstName,
  lastName,
  email,
  dealerName,
  recordType: dealerType
}) => {
  const { mixpanel } = window
  const appAndDeviceProperties = getAppAndDeviceProperties()

  mixpanel.register(appAndDeviceProperties)
  mixpanel.identify(userId)
  mixpanel.people.set(appAndDeviceProperties)
  mixpanel.people.set({
    $first_name: firstName,
    $last_name: lastName,
    $email: email,
    'Dealer Name': dealerName,
    'Dealer Type': capitalizeWord(dealerType)
  })
  mixpanel.track('Find Site', { Found: false })

  return MIXPANEL_EVENT_QUEUED('Find Site - site not found')
}

export const siteFound = (
  {
    uniqueId: userId,
    firstName,
    lastName,
    email,
    dealerName,
    recordType: dealerType
  },
  { city, st_id, postalCode, siteKey, commissioned }
) => {
  const { mixpanel } = window
  const appAndDeviceProperties = getAppAndDeviceProperties()

  mixpanel.register(appAndDeviceProperties)
  mixpanel.identify(userId)
  mixpanel.people.set(appAndDeviceProperties)
  mixpanel.people.set({
    $first_name: firstName,
    $last_name: lastName,
    $email: email,
    'Dealer Name': dealerName,
    'Dealer Type': capitalizeWord(dealerType)
  })
  mixpanel.track('Find Site', {
    Found: true,
    ...{
      $city: city,
      State: st_id,
      'Zip Code': postalCode,
      'Site ID': siteKey,
      Commissioned: commissioned
    }
  })

  return MIXPANEL_EVENT_QUEUED('Find Site - site found')
}

export const saveConfiguration = config => {
  const { mixpanel } = window
  mixpanel.track('Configure', config)
  return MIXPANEL_EVENT_QUEUED('Configure - submit site config')
}

export const saveInventory = (inventory, hasEss) => {
  const { mixpanel } = window

  mixpanel.register({
    'Equinox Storage Site': hasEss
  })

  mixpanel.track('Inventory', {
    'AC Module Count': prop('AC_MODULES', inventory),
    'ESS System Size': hasEss ? prop('ESS', inventory) : 'None',
    'Equinox Storage Site': hasEss
  })
  return MIXPANEL_EVENT_QUEUED('Inventory')
}

export const deviceResume = pageResume => {
  const { mixpanel } = window
  mixpanel.register({
    Page: pageResume
  })
  return MIXPANEL_EVENT_QUEUED('User opens app - Register page resume')
}

/**
 *
 * @param {string}[interfaceType] - The type of connection the user has. either can be  PVS, Cell, or Customer WiFi
 * @return {Action<null, null>}
 */
export const internetConnection = (interfaceType = '') => {
  const { mixpanel } = window
  mixpanel.register({
    'Internet Connection': interfaceType
  })
  return MIXPANEL_EVENT_QUEUED('User opens app - Register page resume')
}

export const firmwareUpdateEvent = data => {
  const { mixpanel } = window
  mixpanel.track('Firmware Update', {
    'Current FW Version': prop('fromFWVersion', data),
    'Current BN Version': prop('fromBuildNumber', data),
    'New FW Version': prop('toFWVersion', data),
    'New BN Version': prop('toBuildNumber', data),
    Success: prop('success', data)
  })
  return MIXPANEL_EVENT_QUEUED('Firmware update - Update successful')
}

export const timeMixPanelEvent = name => {
  if (!name) throw new Error('The name parameter is required')
  const { mixpanel } = window
  mixpanel.time_event(name)
}

export const commissionSite = ({ duration, timeConfiguring }) => {
  const { mixpanel } = window
  mixpanel.track('Commission Site', {
    $duration: duration,
    'Time Configuring': timeConfiguring
  })
  return MIXPANEL_EVENT_QUEUED(
    'Commission Site - first successful site configuration'
  )
}
