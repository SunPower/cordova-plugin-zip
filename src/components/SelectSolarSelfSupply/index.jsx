import React from 'react'
import { useSelector, useDispatch } from 'react-redux'
import ModalSelect from '../ModalSelect'
import { useI18n } from '../../shared/i18n'
import { setSolarSelfSupply } from '../../state/actions/storage'

const createSelfSupplyChangeHandler = dispatch => {
  return option => dispatch(setSolarSelfSupply(option.id))
}

const SelectSolarSelfSupply = ({ disabled }) => {
  const t = useI18n()
  const dispatch = useDispatch()
  const selectedId = useSelector(
    state => state.storage && state.storage.selectedSolarSelfSupplyId
  )
  const options = [
    { id: 1, text: t('ZERO_BACKUP_SOLAR'), textLeft: '0%', value: 0 },
    { id: 2, text: t('TEN_BACKUP_SOLAR'), textLeft: '10%', value: 1 },
    { id: 3, text: t('TWENTY_BACKUP_SOLAR'), textLeft: '20%', value: 2 },
    { id: 4, text: t('THIRTY_BACKUP_SOLAR'), textLeft: '30%', value: 3 },
    { id: 5, text: t('FORTY_BACKUP_SOLAR'), textLeft: '40%', value: 4 },
    { id: 6, text: t('FIFTY_BACKUP_SOLAR'), textLeft: '50%', value: 5 }
  ]
  const selectedOpt = options.find(({ id }) => id === selectedId) || {}
  selectedOpt.selected = true
  const footer = (
    <div className="backup-warning pr-20 pl-20 mb-30">
      <span className="highlighted">{t('BACKUP_MAX')}</span>
      <span>{t('BACKUP_WARNING')}</span>
    </div>
  )
  return (
    <ModalSelect
      header={t('CHOOSE_BACKUP_RESERVE_AMOUNT')}
      options={options}
      className="cost-savings is-seven-eighths"
      footer={footer}
      onClick={createSelfSupplyChangeHandler(dispatch)}
      modalId="modal-solar-self-supply-root"
      disabled={disabled}
    />
  )
}

export default SelectSolarSelfSupply
