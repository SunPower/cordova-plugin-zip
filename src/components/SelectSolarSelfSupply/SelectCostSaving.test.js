import React from 'react'
import { shallow } from 'enzyme'
import * as reactRedux from 'react-redux'
import SelectSolarSelfSupply from '.'
import * as i18n from '../../shared/i18n'

describe('SelectSolarSelfSupply component', () => {
  let dispatchMock

  beforeEach(() => {
    dispatchMock = jest.fn()
    jest
      .spyOn(i18n, 'useI18n')
      .mockImplementation(path => (key, ...params) =>
        `${key.toUpperCase()} ${params.join('_')}`.trim()
      )
    jest.spyOn(reactRedux, 'useDispatch').mockImplementation(() => dispatchMock)
    jest.spyOn(reactRedux, 'useSelector').mockImplementation(() => ({
      storage: {
        selectedCostSavingId: 1
      }
    }))
  })
  it('renders without crashing', () => {
    const component = shallow(<SelectSolarSelfSupply />)
    expect(component).toMatchSnapshot()
  })
})
