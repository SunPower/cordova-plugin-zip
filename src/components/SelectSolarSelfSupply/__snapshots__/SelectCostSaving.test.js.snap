// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectSolarSelfSupply component renders without crashing 1`] = `
<ModalSelect
  className="cost-savings is-seven-eighths"
  footer={
    <div
      className="backup-warning pr-20 pl-20 mb-30"
    >
      <span
        className="highlighted"
      >
        BACKUP_MAX
      </span>
      <span>
        BACKUP_WARNING
      </span>
    </div>
  }
  header="CHOOSE_BACKUP_RESERVE_AMOUNT"
  modalId="modal-solar-self-supply-root"
  onClick={[Function]}
  options={
    Array [
      Object {
        "id": 1,
        "text": "ZERO_BACKUP_SOLAR",
        "textLeft": "0%",
        "value": 0,
      },
      Object {
        "id": 2,
        "text": "TEN_BACKUP_SOLAR",
        "textLeft": "10%",
        "value": 1,
      },
      Object {
        "id": 3,
        "text": "TWENTY_BACKUP_SOLAR",
        "textLeft": "20%",
        "value": 2,
      },
      Object {
        "id": 4,
        "text": "THIRTY_BACKUP_SOLAR",
        "textLeft": "30%",
        "value": 3,
      },
      Object {
        "id": 5,
        "text": "FORTY_BACKUP_SOLAR",
        "textLeft": "40%",
        "value": 4,
      },
      Object {
        "id": 6,
        "text": "FIFTY_BACKUP_SOLAR",
        "textLeft": "50%",
        "value": 5,
      },
    ]
  }
/>
`;
